### ISSUE TITLE
[RFE] Consider tooling for easily spotting dependencies

### ISSUE THREAD

**Issue by Keruspe** [mesonbuild/meson#9541]
When meson brought a separate `meson_options.txt`  it made things a lot easier for packagers to see what options are available, and we really have to thank you for that!

On the other hand, wrt system dependencies, it's another story, as we enter the same tragic (for packagers!) scenario as with CMake. Basically, people end up with a ton of `meson.build` files across their project hierarchy and each one of those can list new dependencies.

Autotools were very far from perfect, and meson is a very nice improvement over them, but the one thing they (nearly) got right was this: apart from a few exceptions (looking at you, custom m4 macros), all the dependencies were listed in one file.

Having something like `meson_dependencies.txt` for this, or a tool such as `meson packaging-report` which would list the dependencies and their conditions (e.g. "only when option foo is enabled") would be really nice, and a "never seen before" feature that would make meson *the best* build tool for packagers too, when it seem to already be the one for developers.

**Comment by eli-schwartz** [comment]
> Autotools were very far from perfect, and meson is a very nice improvement over them, but the one thing they (nearly) got right was this: apart from a few exceptions (looking at you, custom m4 macros), all the dependencies were listed in one file.

*apart from a few exceptions*

*cries*

Moving on from there, the fact that you could often find all dependencies listed in configure.ac was a side effect of the fact that all configuration logic had to be in that one file (of course, only until m4 combinatorial explosion). It is of course perfectly possible for a single meson.build file to contain the entire build system, without use of `subdir()`, in which case all the dependencies are once more listed in one file.

Adding a meson_dependencies.txt would definitely be quite new as far as build systems go, but then again, the same goal could be achieved by convincing projects to always search for all dependencies as the first thing in the root meson.build -- and the conditions for dependencies could be quite complicated, it's not just when given options are enabled, Options might depend on things like which OS and compiler the build is configured for, or on whether or not some other dependency was found, or various other things that might be hard to express in a flat dependency manifest.

I wonder if you have any thoughts on resolving that sort of conditional in a flat dependency manifest? :)

**Comment by Keruspe** [comment]
> _apart from a few exceptions_
> 
> _cries_
> 

That's just my experience as a packager, maybe I got lucky. Anyways...

> Moving on from there, the fact that you could often find all dependencies listed in configure.ac was a side effect of the fact that all configuration logic had to be in that one file (of course, only until m4 combinatorial explosion). It is of course perfectly possible for a single meson.build file to contain the entire build system, without use of `subdir()`, in which case all the dependencies are once more listed in one file.

Yes, but that definitely doesn't seem to be the goal of meson, right?

> I wonder if you have any thoughts on resolving that sort of conditional in a flat dependency manifest? :)

I'm well aware that this is not a trivial matter and would probably involve some dark magic, which is why I also had in mind some packager-oriented command as an alternative.

For dependencies to be declared in a flat file, a simple (yet suboptimal) option could be to list everything there while considering everything to be `optional: true`, then in the meson.build, to use e.g. `foobar_dep` as before for optional dependencies, and add something like `foobar_dep.required()` to assert that the dep was found for non-optional ones.

Something like `meson packaging-report` would probably be a better fit, but I'm not familiar enough with meson internals to tell if that'd be doable or not, something that would output stuff like this:

```
pkg-config dependencies:
foo-1.0 >= 1.3
bar-2 if bar=enabled
baz if arch=x86_64

tools required:
perl
md2doc 
```

**Comment by mensinda** [comment]
We do have an [introspection API](https://mesonbuild.com/IDE-integration.html) where you can list all the dependencies. It should also work (via the programmatic interface) without a configured build directory.

**Comment by eli-schwartz** [comment]
`meson introspect` is totally useless without a configured build directory, since it will always return:

```
Current directory is not a meson build directory.
Please specify a valid build dir or change the working directory to it.
```

and at that point you might as well use the json files themselves.

It keeps on being a really annoying limitation, especially for things like `--projectinfo` where you only want to parse the very first `project()` definition, no variables are feasible, and you could even do it with sed except that you have no guarantee of where whitespace/linebreaks are.

**Comment by mensinda** [comment]
I guess it is not really documented. It works like this: `meson introspect --dependencies meson.build`. This works without a build directory. The trick here is to supply the path to a `meson.build` instead of a build directory.

**Comment by eli-schwartz** [comment]
Oh wow, yeah, I was totally unaware of that. :D Note that it appears to be tricky to perfectly deduce, since it works for some projects but does not work for muon which uses a foreach loop:

```
foreach d: [
		['libcurl',    'libcurl',    'dependency'],
		['libarchive', 'libarchive', 'dependency'],
		['libpkgconf', 'libpkgconf', 'dependency'],
		['samurai',    'samu',       'subproject', 'libsamu_dep'],
	]
		dep = dependency(
			d[0],
			required: get_option(d[1]),
			static: get_option('static')
		)
		found = dep.found()
endforeach
```

Ironically, this foreach loop seems to be an attempt at declaratively defining dependencies. :D

**Comment by mensinda** [comment]
> Note that it appears to be tricky to perfectly deduce, since it works for some projects but does not work for muon which uses a foreach loop

Yes, this is a known limitation, since the actual interpreter does not run and the results are more of a best effort. Branches also don't work (all branches are evaluated, so you get the dependencies for both the `if` and `else` branch).

**Comment by Keruspe** [comment]
This is nice indeed, will already help a lot
